public/js/utils/utils.js

```diff
@@ -62,18 +61,29 @@ function endTruncateStr(str: any, size: any) {
   return str;
 }
 
-function workerTask(worker: any, message: any) {
-  let deferred = defer();
-  worker.postMessage(message);
-  worker.onmessage = function(result) {
-    if (result.data && result.data.error) {
-      deferred.reject(result.data.error);
-    }
-
-    deferred.resolve(result.data);
+let msgId = 1;
+function workerTask(worker: any, method: string) {
+  return function(...args) {
+    return new Promise((resolve, reject) => {
+      const id = msgId++;
+      worker.postMessage({ id, method, args });
+
+      const listener = ({ data: result }) => {
+        if (result.id !== id) {
+          return;
+        }
+
+        worker.removeEventListener("message", listener);
+        if (result.error) {
+          reject(result.error);
+        } else {
+          resolve(result.response);
+        }
+      };
+
+      worker.addEventListener("message", listener);
+    });
```

Comment:
Yes, messages may not come back in the same order. If we want a shared util for workers, we should assume that message might not be ordered. I suppose if the worker is fully synchronous, then messages will be responded to in the same order (hadn't really realized that, good point), but now that might do fetching in the worker it's not guaranteed anymore.

We did a similar thing in mozilla-central: https://github.com/mozilla/gecko-dev/blob/master/devtools/shared/worker/worker.js#L71

You're right though that it would be fine if all the functions are synchronous. I did base a lot of this on how you did it before, so good work.